pkg/storage/pod.go

```diff
@@ -101,21 +106,59 @@ func (s *podStorage) Store(newPods []PodMetricsPoint) {
 		prevContainers := make(map[string]ContainerMetricsPoint, len(newPod.Containers))
 		for _, newContainer := range newPod.Containers {
 			if _, exists := lastContainers[newContainer.Name]; exists {
-				klog.ErrorS(nil, "Got duplicate Container point", "container", newContainer.Name, "pod", klog.KRef(newPod.Namespace, newPod.Name))
+				klog.ErrorS(nil, "Got duplicate container point",
+					"container", newContainer.Name,
+					"pod", klog.KRef(newPod.Namespace, newPod.Name))
 				continue
 			}
+
+			lastPod, foundLastPod := s.last[podRef]
+			var foundLastCont = false
+			var lastContainer = ContainerMetricsPoint{}
+			if foundLastPod {
+				lastContainer, foundLastCont = lastPod[newContainer.Name]
+			}
+
+			prevPod, foundPrevPod := s.prev[podRef]
+			var foundPrevCont = false
+			var prevContainer = ContainerMetricsPoint{}
+			if foundPrevPod {
+				prevContainer, foundPrevCont = prevPod[newContainer.Name]
+			}
+
+			// Got older data point, just drop it.
+			if foundLastCont && newContainer.Timestamp.Before(lastContainer.Timestamp) {
+				lastContainers[newContainer.Name] = lastContainer
+				if foundPrevCont {
+					prevContainers[newContainer.Name] = prevContainer
+				}
+
+				klog.InfoS("Got older node metrics point, drop it",
+					"container", newContainer.Name,
+					"pod", klog.KRef(newPod.Namespace, newPod.Name),
+					"lastTimestamp", lastContainer.Timestamp,
+					"timestamp", newContainer.Timestamp)
+				continue
+			}
```

Comment:
This changes the default behavior of storage to always accept latest sample. Instead of changing this behavior could you just check if new point is older then prev point, if so just don't store prev point.

Code:
```
if lastPod, found := s.last[podRef]; found {
				// Keep previous metric point if newContainer has not restarted (new metric start time < stored timestamp)
				if lastContainer, found := lastPod[newContainer.Name]; found && newContainer.StartTime.Before(lastContainer.Timestamp) {
					// If new point is different then one already stored
					if newContainer.Timestamp.After(lastContainer.Timestamp) {
						// Move stored point to previous
						prevContainers[newContainer.Name] = lastContainer
					} else if prevPod, found := s.prev[podRef]; found && prevPod[newContainer.Name].Timestamp.Before(newContainer.Timestamp) { //<<< THe difference
						// Keep previous point
						prevContainers[newContainer.Name] = prevPod[newContainer.Name]
					}
				}
			}
```

Reply:
Sorry about that. I change the behavior according to the comments from @dgrisonnet, https://github.com/kubernetes-sigs/metrics-server/issues/750#issuecomment-840108196 .

>  but for the storing part shouldn't we just skip these obsolete samples instead of taking them into account?

I thought it is reasonable to drop older data point. And I have restore code here.